### PR TITLE
Restrict opening engine number

### PR DESCRIPTION
**Pull request by DorianZheng**
Signed-off-by: DorianZheng <xingzhengde72@gmail.com>

## What have you changed? (mandatory)
Restrict opening engine number. Each opening engine will occupy large memory such that OOM will take place occasionally if we don't restrict opening engine number.

## What are the type of the changes? (mandatory)
Improvement

## How has this PR been tested? (mandatory)
make trace_test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

### CODE REVIEW

**Comment by breeswish** [comment]
Could you list reasons for this change in the description?

File: src/import/kv_importer.rs [hunk 1]
```diff
@@ -59,6 +59,13 @@ impl KVImporter {
             return Ok(());
         }
 
+        // Restrict opening engine num
+        if inner.engines.len() >= self.cfg.num_engines {
+            let errmsg = format!("Too much opening engines {}: {}", uuid, inner.engines.len());
+            error!("{}", errmsg);
+            return Err(Error::RocksDB(errmsg));
```
Comment (huachaohuang) [inline src/import/kv_importer.rs]: This is not a RocksDB error, please define a more specific error.

File: etc/tikv-importer.toml [hunk 2]
```diff
@@ -49,3 +49,5 @@ num-import-jobs = 24
 # region-split-size = "96MB"
 # stream channel window size, stream will be blocked on channel full.
 # stream-channel-window = 128
+# number of opening engines
+num-engines = 6
```
Comment (huachaohuang) [inline etc/tikv-importer.toml]: I think `max-engines` is better.

Comment (siddontang) [reply etc/tikv-importer.toml]: what dos engine mean?

Comment (DorianZheng) [reply etc/tikv-importer.toml]: engine means storage engine

File: src/import/kv_importer.rs (hunk 1 above)
Comment (DorianZheng) [reply src/import/kv_importer.rs]: Sure

**Comment by DorianZheng** [comment]
@huachaohuang @zhangjinpeng1987 PTAL

File: etc/tikv-importer.toml [hunk 3]
```diff
@@ -49,3 +49,5 @@ num-import-jobs = 24
 # region-split-size = "96MB"
 # stream channel window size, stream will be blocked on channel full.
 # stream-channel-window = 128
+# number of opening engines
+max-storage-engines = 6
```
Comment (huachaohuang) [inline etc/tikv-importer.toml]: max-open-engines

File: etc/tikv-importer.toml [hunk 4]
```diff
@@ -49,3 +49,5 @@ num-import-jobs = 24
 # region-split-size = "96MB"
 # stream channel window size, stream will be blocked on channel full.
 # stream-channel-window = 128
+# number of opening engines
```
Comment (huachaohuang) [inline etc/tikv-importer.toml]: maximum number of open engines

File: src/import/config.rs [hunk 5]
```diff
@@ -61,6 +63,9 @@ impl Config {
         if self.stream_channel_window == 0 {
             return Err("import.stream_channel_window can not be 0".into());
         }
+        if self.max_storage_engines == 0 {
+            return Err("import.num_engines can not be 0".into());
```
Comment (huachaohuang) [inline src/import/config.rs]: Update the error message

File: src/import/errors.rs [hunk 6]
```diff
@@ -109,6 +109,9 @@ quick_error! {
         PrepareRangeJobFailed(tag: String) {
             display("{}", tag)
         }
+        ResourceTemporarilyUnavailable(tag: String) {
```
Comment (huachaohuang) [inline src/import/errors.rs]: s/tag/msg

File: src/import/kv_importer.rs [hunk 7]
```diff
@@ -59,6 +59,13 @@ impl KVImporter {
             return Ok(());
         }
 
+        // Restrict opening engine num
+        if inner.engines.len() >= self.cfg.max_storage_engines {
+            let errmsg = format!("Too much opening engines {}: {}", uuid, inner.engines.len());
```
Comment (huachaohuang) [inline src/import/kv_importer.rs]: "Too many open engines ..."

File: src/import/kv_importer.rs [hunk 8]
```diff
@@ -59,6 +59,13 @@ impl KVImporter {
             return Ok(());
         }
 
+        // Restrict opening engine num
```
Comment (huachaohuang) [inline src/import/kv_importer.rs]: "Restrict max open engines"

File: etc/tikv-importer.toml [hunk 9]
```diff
@@ -49,3 +49,5 @@ num-import-jobs = 24
 # region-split-size = "96MB"
 # stream channel window size, stream will be blocked on channel full.
 # stream-channel-window = 128
+# maximum number of open engines
+max-open-engines = 6
```
Comment (kennytm) [inline etc/tikv-importer.toml]: Could the max-open-engines be increased to 8? Or should we reduce the table-concurrency in lightning to 6?

cc @holys

**Comment by DorianZheng** [comment]
@zhangjinpeng1987 PTAL

**Pull request merged**